test/test_utilities.cpp

```diff
@@ -48,6 +48,18 @@ void AssertTensorEquals(Tensor& T1, Tensor& T2) {
   }
 }
 
+void AssertTensorEqualsInt32(Tensor& T1, Tensor& T2) {
+  ASSERT_EQ(T1.shape(), T2.shape());
+  auto T_size = T1.flat<int32>().size();
+  auto T1_data = T1.flat<int32>().data();
+  auto T2_data = T2.flat<int32>().data();
+  for (int k = 0; k < T_size; k++) {
+    auto a = T1_data[k];
+    auto b = T2_data[k];
+    EXPECT_EQ(a, b);
+  }
+}
+
```

Comment:
We can (somewhat) templatize AssertTensorEqual as follows:
We pass the definition of equality to AssertTensorEqual and  use EXPECT_TRUE, instead of relying on EXPECT_EQUAL or EXPECT_FLOAT_EQ.
```
template <typename T>
void AssertTensorEqual(Tensor& T1, Tensor& T2, std::function<bool(T, T)> eq){
    ASSERT_EQ(T1.shape(), T2.shape());
    auto T_size = T1.flat<T>().size();
   auto T1_data = T1.flat<T>().data();
   auto T2_data = T2.flat<T>().data();
  for (int k = 0; k < T_size; k++) {
    auto a = T1_data[k];
    auto b = T2_data[k];
    EXPECT_TRUE(eq(a, b));
  }
}
```

Reply:
```eq``` can have a default value:
```void AssertTensorEqual(Tensor& T1, Tensor& T2, std::function<bool(T, T)> eq = [](T arg0, T arg1){return arg0==arg1})```

This would take care of bool and int. so the switchcase will be much reduced now:
```
switch (data_type_enum){
case dt_int:
case dt_bool:
case others_where_vanilla_equality_suffices:
  AssertTensorEqual(a, b)
  break;
case dt_float:
  AssertTensorEqual(a, b, custom_eq_op)
default:
  fail here
}
```

or, we could have the default case as vanilla equality

Reply:
Thanks @sayantan-nervan.  @kanvi-nervana  and I had been discussing how to templatize/overload this function. This is a good option. Most of our ops have float outputs.  So , need to figure out how the float comparison would work out with this, we might have to change the function to take in tolerance std::function<bool(T, T, tolerance)> eq

Reply:
here's another possibility:
we have the AssertTensorEqual templated on T
```
template <typename T>
AssertTensorEqual(Tensor& T1, Tensor& T2){
}
```

Then we have:
```
EXPECT_TRUE(eq<T,T>(a, b))
```

Finally we have a template specialization:
```
template <typename T>
bool eq(T arg0, T arg1){ return arg0==arg1;}

template <>
bool eq(float arg0, float arg1){return arg0 == arg1;}
```

Now this removes the requirement of even the switch case

Reply:
Yes, this is even better! 👍

Reply:
Another consideration, here would be the error messages. On using EXPECT_EQ()/ EXPECT_FLOAT_EQ() When the test case fails, the message prints both the values we are comparing (a and b) .  EXPECT_TRUE() would print out the boolean value returned by the function eq and compare it with true. We wont see the actual values being compared, I think. So there is a case for using EXPECT_EQ and EXPECT_FLOAT_EQ .

Reply:
I haven't implemented the latest suggestion but if we want to consider that we could add debug logs.